connectd/connectd.c

```diff
@@ -1097,28 +1128,71 @@ static struct wireaddr_internal *setup_listeners(const tal_t *ctx,
 	for (size_t i = 0; i < tal_count(proposed_wireaddr); i++) {
 		if (!(proposed_listen_announce[i] & ADDR_LISTEN))
 			continue;
-
 		if (proposed_wireaddr[i].itype != ADDR_INTERNAL_AUTOTOR)
 			continue;
 		if (!(proposed_listen_announce[i] & ADDR_ANNOUNCE)) {
 				tor_autoservice(tmpctx,
-						&proposed_wireaddr[i].u.torservice,
+						&proposed_wireaddr[i].u.torservice.torservice_address,
 						tor_password,
 						binding,
 						daemon->use_v3_autotor);
 			continue;
 		};
 		add_announcable(announcable,
 				tor_autoservice(tmpctx,
-						&proposed_wireaddr[i].u.torservice,
+						&proposed_wireaddr[i].u.torservice.torservice_address,
 						tor_password,
 						binding,
 						daemon->use_v3_autotor));
 	}
 
+	/* Now we have bindings, set up any Tor static addresses: we will point
+	 * it at the first bound IPv4 or IPv6 address we have. */
+	for (size_t i = 0; i < tal_count(proposed_wireaddr); i++) {
+		if (!(proposed_listen_announce[i] & ADDR_LISTEN))
+			continue;
+		if (proposed_wireaddr[i].itype != ADDR_INTERNAL_STATICTOR)
+			continue;
+		blob = proposed_wireaddr[i].u.torservice.blob;
+
+		if (pubkey_from_node_id(&pb, &daemon->id)) {
+			if (strstr(proposed_wireaddr[i].u.torservice.blob, STATIC_TOR_MAGIC_STRING)) {
+				if (sodium_mlock(&random, sizeof(random)) != 0)
+						status_failed(STATUS_FAIL_INTERNAL_ERROR,
+									"Could not lock the random prf key memory.");
+				randombytes_buf((void * const)&random, 32);
+				/* generate static tor node address, take first 32 bytes from secret of node_id plus 32 random bytes from sodiom */
+				struct sha256 sha;
+				sha256(&sha, hsm_do_ecdh(tmpctx, &pb), 32);
+				/* even if it's a partly secret, tor shall see only the sha */
+				blob = tal_fmt(tmpctx, "%.32s%.32s", tal_hexstr(tmpctx, &sha, 32), tal_hexstr(tmpctx, &random, 32));
+			}
```

Comment:
sodium_munlock(&random) is needed at the end.  Which also zeroes.

Seems like random can be a local var here, too?  As can pb...

Reply:
done. changed.